# <font color=darkred>Laboratory 7: FUNctions </font>

In [17]:
# Preamble script block to identify host, user, and kernel
import sys
! hostname
! whoami
print(sys.executable)
print(sys.version)
print(sys.version_info)

GouldPC
gouldpc\gould
C:\Users\gould\anaconda3\python.exe
3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 15:03:56) [MSC v.1929 64 bit (AMD64)]
sys.version_info(major=3, minor=12, micro=4, releaselevel='final', serial=0)


## Full name: Graysen Gould
## R#: 11944656
## Title of the notebook: Laboratory 7: FUNctions
## Date: 9-23-24
___

![](https://runestone.academy/runestone/books/published/fopp/_images/function_call.gif) <br>

## <font color=purple>What is a function in Python?</font>

Functions are simply pre-written code fragments that perform a certain task.
In older procedural languages functions and subroutines are similar, but a function returns a value whereas
a subroutine operates on data. 
The difference is subtle but important. 

More recent thinking has functions being able to operate on data (they always could) and the value returned may be simply an exit code.
An analogy are the functions in *MS Excel*. 
To add numbers, we can use the sum(range) function and type `=sum(A1:A5)` instead of typing `=A1+A2+A3+A4+A5`
___
## <font color=purple>Calling the Function</font>

We call a function simply by typing the name of the function or by using the dot notation.
Whether we can use the dot notation or not depends on how the function is written, whether it is part of a class, and how it is imported into a program.

Some functions expect us to pass data to them to perform their tasks. 
These data are known as parameters( older terminology is arguments, or argument list) and we pass them to the function by enclosing their values in parenthesis ( ) separated by commas. 

For instance, the `print()` function for displaying text on the screen is \called" by typing `print('Hello World')` where print is the name of the function and the literal (a string) 'Hello World' is the argument.
___
## <font color=purple>Program flow</font>

A function, whether built-in, or added must be defined *before* it is called, otherwise the script will fail.  Certain built-in functions "self define" upon start (such as `print()` and `type()` and we need not worry about those funtions).

___

___
## <font color=purple>Built-In in Primitive Python (Base install)</font>

The base Python functions and types built into it that are always available, the figure below lists those functions.

![](https://itvoyagers.in/wp-content/uploads/2020/09/built-in-functions-itvoyagers.in_.png)

Notice all have the structure of `function_name()`, except `__import__()` which has a constructor type structure, and is not intended for routine use.  We will learn about constructors later.


___
## <font color=purple>Added-In using External Packages/Modules and Libraries (e.g. math)</font>

Python is also distributed with a large number of external functions. 
These functions are saved
in files known as modules. 
To use the built-in codes in Python modules, we have to import
them into our programs first. We do that by using the import keyword. 
There are three
ways to import:
1. Import the entire module by writing import moduleName; For instance, to import the random module, we write import random. To use the randrange() function in the random module, we write random.randrange( 1, 10);28
2. Import and rename the module by writing import random as r (where r is any name of your choice). Now to use the randrange() function, you simply write r.randrange(1, 10); and
3. Import specific functions from the module by writing from moduleName import name1[,name2[, ... nameN]]. For instance, to import the randrange() function from the random module, we write from random import randrange. To import multiple functions, we separate them with a comma. To import the randrange() and randint() functions, we write from random import randrange, randint. To use the function now, we do not have to use the dot notation anymore. Just write randrange( 1, 10).

An example below  will illustrate, change the cell to code and run it, you should get an error.
Then fix the indicated line (remove the leading "#" in the import math ... line) and rerun, should get a functioning script.

### Example: Import Square Root

In [18]:
# reset the notebook using a magic function in JupyterLab
%reset -f 

In [19]:
# An example, run once as is then activate indicated line, run again - what happens?
x = 4.
sqrt_by_arithmetic = x**0.5
print('Using arithmetic square root of ', x, ' is ',sqrt_by_arithmetic )

Using arithmetic square root of  4.0  is  2.0


In [20]:
import math # import the math package ## activate and rerun

sqrt_by_math = math.sqrt(x)  # note the dot notation
print('Using math package square root of ',  x,' is ',sqrt_by_arithmetic)

Using math package square root of  4.0  is  2.0


In [21]:
# Here is an alternative way: We just load the function that we want:
# reset the notebook using a magic function in JupyterLab

%reset -f 

# An example, run once as is then activate indicated line, run again - what happens?
x= 4.
sqrt_by_arithmetic = x**0.5
print('Using arithmetic square root of ', x, ' is ',sqrt_by_arithmetic )


from math import sqrt # import sqrt from the math package ## activate and rerun
sqrt_by_math = sqrt(x)  # note the notation
print('Using math package square root of ',  x,' is ',sqrt_by_arithmetic)

Using arithmetic square root of  4.0  is  2.0
Using math package square root of  4.0  is  2.0


### Example 2: Import randrange

In [22]:
# Example 1 of import
%reset -f 

import random

low = 1 ; high = 10
random.randrange(low,high) #generate random number in range low to high

2

In [23]:
# Example 2 of import
%reset -f 

import random as r

low = 1 ; high = 10
r.randrange(low,high)

8

In [24]:
# Example 3 of import

%reset -f 

from random import randrange 

low = 1 ; high = 10
randrange(low,high)



6

___
The modules that come with Python are extensive and listed at 
https://docs.python.org/3/py-modindex.html.
There are also other modules that can be downloaded and used
(just like user defined modules below). 
In these labs we are building primitive codes to learn how to code and how to create algorithms. 
For many practical cases you will want to load a well-tested package to accomplish the tasks. 

That exercise is saved for the end of the document.

## <font color=purple>User-Built</font>

We can define our own functions in Python and reuse them throughout the program.
The syntax for defining a function is:

    def functionName( argument ):
        code detailing what the function should do
        note the colon above and indentation
        ...
        ...
        return [expression]
        
The keyword `def` tells the program that the indented code from the next line onwards is
part of the function. 
The keyword `return `tells the program to return an answer from the
function. 
There can be multiple return statements in a function. 
Once the function executes
a return statement, the program exits the function and continues with *its* next executable
statement. 
If the function does not need to return any value, you can omit the return
statement.

Functions can be pretty elaborate; they can search for things in a list, determine variable
types, open and close files, read and write to files. 

To get started we will build a few really
simple mathematical functions; we will need this skill in the future anyway, especially in
scientific programming contexts.

___
### <font color=purple>User-built within a Code Block</font>

### The Foo Function
 
For our first function we will code $$f(x) = x\sqrt{1 + x}$$ into a function named `foo()`.

When you run the next cell, all it does is prototype the function (defines it), nothing happens until we use the function.

In [28]:
def foo(x) :
    temp = x * ((1.0+x)**(0.5)) # don't need the math package
    return temp
# the function should make the evaluation
# store in the local variable temp
# return contents of temp

In [29]:
# wrapper to run the foo function
yes = 0
while yes == 0:
    xvalue = input('enter a numeric value')
    try:
        xvalue = float(xvalue)
        yes = 1
    except:
        print('enter a bloody number! Try again \n')
        
# call the function, get value , write output
yvalue = foo(xvalue)
print('f(',xvalue,') = ',yvalue) # and we are done 

enter a numeric value e


enter a bloody number! Try again 



enter a numeric value e


enter a bloody number! Try again 



enter a numeric value 3


f( 3.0 ) =  6.0


___
### The 3 Number Average Function

Create the AVERAGE function for three values and test it for these values:
- 3,4,5
- 10,100,1000
- -5,15,5

In [30]:
def AVERAGE3(x,y,z) : #define the function "AVERAGE3"
    Ave = (x+y+z)/3   #computes the average
    return Ave

In [31]:
print(AVERAGE3(3,4,5))

4.0


In [32]:
var1 = AVERAGE3(10,100,1000)
typeVar1 = type(var1)

print(typeVar1, ' | value:', var1)

<class 'float'>  | value: 370.0


In [33]:
print(AVERAGE3(-5,15,5))

5.0


___
### The KATANA Function

Create the Katana function for rounding off to the nearest hundredths (to 2 decimal places) and test it for these values:
- 25.33694
- 15.753951
- 3.14159265359

In [34]:
def Katana(x) : #define the function "Katana"
    newX = round(x, 2)
    return newX

In [35]:
print(Katana(25.33694))

25.34


In [36]:
print(Katana(15.753951))

15.75


In [37]:
print(Katana(3.14159265359))

3.14


___
## <font color=purple>Variable Scope</font>
An important concept when defining a function is the concept of variable scope. 
Variables defined inside a function are treated differently from variables defined outside. 
Firstly, any variable declared within a function is only accessible within the function. 
These are known as local variables. 

In the `foo()` function, the variables `x` and `temp` are local to the function.
Any variable declared outside a function in a main program is known as a program variable
and is accessible anywhere in the program. 

In the example, the variables `xvalue` and `yvalue` are program variables (global to the program; if they are addressed within a function, they could be operated on.)
Generally we want to protect the program variables from the function unless the intent is to change their values. 
The way the function is written in the example, the function cannot damage `xvalue` or `yvalue`.

If a local variable shares the same name as a program variable, any code inside the function is
accessing the local variable. Any code outside is accessing the program variable

## Example 1
Write a function to calculate the area of a circle

#### With no Parameter and no return value

In [79]:
def calc_area_1 ():
    r = float(input("please input the radius of a circle"))
    a = round(3.1415*(r**2), 2)
    print(f"the area of a circle of radius {r} is {a}")
calc_area_1()

please input the radius of a circle 3


the area of a circle of radius 3.0 is 28.27


#### With no Parameter and return value

In [80]:
def calc_area_2 ():
    r = float(input("please input the radius of a circle"))
    return round(3.1415*(r**2), 2)
    print(f"the area of a circle of radius {r} is {a}")
area = calc_area_2()

print(f"the area of the circle is {area}")

please input the radius of a circle 3


the area of the circle is 28.27


#### With Parameter and no return value

In [81]:
def calc_area_3 (r):
    a = round(3.1415*(r**2), 2)
    print(f"the area of a circle of radius {r} is {a}")
    
r = float(input("please input the radius of a circle"))
calc_area_3(r)

please input the radius of a circle 3


the area of a circle of radius 3.0 is 28.27


#### With Parameter and return value

In [47]:
def calc_area_4 (r):
    return round(3.1415*(r**2), 2)

In [82]:
r = float(input("please input the radius of a circle"))
area = calc_area_4(r)
print(f"A cirlce with a radius of {r} has an area = {area}")

please input the radius of a circle 3


A cirlce with a radius of 3.0 has an area = 28.27


## Example 2
Write a function that adds all the digits of an integer. if the resulting sum is more than one digit keep repeating till the sum is  one digit

In [74]:
def add_digits (num):
    sum = 0
    for i in str(num):
        sum += int(i)
    if sum >= 10:
        return add_digits(sum) # recursive funtionality
    else:
        return sum

In [77]:
my_int = 123456789
sum = add_digits(my_int)

print(f"the sum of the digits (done until only one digit) of {my_int} are {sum} ")

#first round 123456789 -> 45
#second round 45 -> 9
# function returns 0

the sum of the digits (done until only one digit) of 123456789 are 9 


___
### <font color=purple>As Separate Module/File</font>

In this section we will use the functions we created earlier as an external Module by storing the functions in a File using the following steps- 

1. Create a text file named "mylibrary.txt"
2. Copy the functions script below into that file.

       def foo(x) :
        temp = x * ((1.0+x)**(0.5)) # don't need the math package
        return temp

       def AVERAGE3(x,y,z) : #define the function "AVERAGE3"
        Ave = (x+y+z)/3   #computes the average
        return Ave

        def Katana(x) : #define the function "Katana"
        newX = round(x, 2)
        return newX

4. rename mylibrary.txt to mylibrary.py
5. modify the wrapper script to use the functions as an external module

In [45]:
import library

In [46]:
print(library.foo(1))

1.4142135623730951


In [47]:
print(library.AVERAGE3(1,2,3))

2.0


In [48]:
print(library.Katana(3.1415))

3.14


## EXERCISE 7.1 (80 marks)
1. We would like to have a program that asks user for three numbers. We want this program to calculate the arithmetic mean of these three numbers, then display the result on the console.<br>
Please calculate the mean of these three numbers using a ***function*** named `avgFunc`. 




In [8]:
# add a comment to every line of code to explain what you are doing,

In [6]:
def avgFunc(): # inputs a list of numbers
    nums = [] #create empty list 
    nums.append(float(input("please input the first number"))) # get first number 
    nums.append(float(input("please input the second number"))) # get second number 
    nums.append(float(input("please input the third number")))# get third number 
    print(f"Your three numbers are {nums}") # output numbers to console
    return round(sum(nums) / len(nums),3) # return the average of three numbers 

avg = avgFunc() # call function and take output 

print(f"The average of these numbers is {avg}") # return average of numbers to user

please input the first number 1
please input the second number 2
please input the third number 3


Your three numbers are [1.0, 2.0, 3.0]
The average of these numbers is 2.0


In [12]:
# test your function with these three sets of inputs: {0, 3, 5}, {-5, 5, 10}, {2.5, 7, 13},

In [10]:
avg = avgFunc()
print(f"The average of these numbers is {avg}")

please input the first number 0
please input the second number 3
please input the third number 5


Your three numbers are [0.0, 3.0, 5.0]
The average of these numbers is 2.667


In [11]:
avg = avgFunc()
print(f"The average of these numbers is {avg}")

please input the first number -5
please input the second number 5
please input the third number 10


Your three numbers are [-5.0, 5.0, 10.0]
The average of these numbers is 3.333


In [12]:
avg = avgFunc()
print(f"The average of these numbers is {avg}")

please input the first number 2.5
please input the second number 7
please input the third number 13


Your three numbers are [2.5, 7.0, 13.0]
The average of these numbers is 7.5


In [ ]:
# report the results of your tests in a markdown cell below.

#### First test:
please input the first number 0 <br>
please input the second number 3 <br>
please input the third number 5 <br>
Your three numbers are [0.0, 3.0, 5.0] <br>
The average of these numbers is 2

#### Second test:
please input the first number -5 <br>
please input the second number 5 <br>
please input the third number 10 <br>
Your three numbers are [-5.0, 5.0, 10.0] <br>
The average of these numbers is 3.333

#### Third test:
please input the first number 2.5 <br>
please input the second number 7 <br>
please input the third number 13 <br>
Your three numbers are [2.5, 7.0, 13.0] <br>
The average of tese numbers is 7.5
 7.5.667

2. write a funtion that prints a pattern of stars using nested for loops:

In [49]:
def pattern (n):
    for i in range(n):
        for j in range(n -i):
            print(" ", end = "")
        for j in range(i):
            print("* ", end = "")
        print("")

pattern(10)

          
         * 
        * * 
       * * * 
      * * * * 
     * * * * * 
    * * * * * * 
   * * * * * * * 
  * * * * * * * * 
 * * * * * * * * * 


3. write a function that calculate the sum of numbers from a start value to an end value using a loop:

In [50]:
def summation (start, end):
    total = 0
    for i in range(start, end + 1):
        total += i
    return total


print(summation(5,10))
    

45


4. write a function that lists all prime numbers from a start value to an end value using a loop:

In [43]:
def get_primes (start, end):
    primes = []
    for i in range(start, end + 1):
        prime = True
        for j in range(2,i):
            if i%j == 0:
                prime = False
        if prime:
            primes.append(i)
    return primes
                

print(get_primes(10,20))

[11, 13, 17, 19]


---
## Lab Exercise 7.2 (20 marks)
Ensure you entered your full name and R Number in the top cell<br>
Convert to PDF and submit both PDF and IPYNB


___
![](https://media2.giphy.com/media/5nj4ZZWl6QwneEaBX4/source.gif) <br>


*Here are some great reads on this topic:* 
- __"Functions in Python"__ available at *https://www.geeksforgeeks.org/functions-in-python/<br>
- __"Defining Your Own Python Function"__ by __John Sturtz__ available at *https://realpython.com/defining-your-own-python-function/ <br>
- __"Graph Plotting in Python | Set 1"__ available at *https://www.geeksforgeeks.org/graph-plotting-in-python-set-1/ <br>
- __"Python Plotting With Matplotlib (Guide)"__ by __Brad Solomon__ available at *https://realpython.com/python-matplotlib-guide/ <br>

*Here are some great videos on these topics:* 
- __"How To Use Functions In Python (Python Tutorial #3)"__ by __CS Dojo__ available at *https://www.youtube.com/watch?v=NSbOtYzIQI0 <br>
- __"Python Tutorial for Beginners 8: Functions"__ by __Corey Schafer__ available at *https://www.youtube.com/watch?v=9Os0o3wzS_I <br>
- __"Python 3 Programming Tutorial - Functions"__ by __sentdex__ available at *https://www.youtube.com/watch?v=owglNL1KQf0 <br><br>

![](https://media.csesoc.org.au/content/images/2019/10/learn11.gif) <br>
___
